### PR TITLE
[WIP] Fail fast on neutral failures

### PR DESCRIPTION
**Pull request by backus**
This PR introduces an incremental improvement in failing fast on neutral failures. This is my first pass at implementing neutral failure fast failing. I still need to add functionality for displaying a special error message in cases where execution was halted due to a neutral failure. Otherwise this code is ready for review

### CODE REVIEW

File: spec/unit/mutant/result/subject_spec.rb [hunk 1]
```diff
@@ -106,4 +106,30 @@
...
   end
+
+  describe '#neutral_failure?' do
+    subject { object.neutral_failure? }
+
+    context 'without failing neutral mutations' do
+      let(:mutation_results) do
+        [
+          instance_double(Mutant::Result::Mutation, success?: false, neutral_failure?: false),
+          instance_double(Mutant::Result::Mutation, success?: true, neutral_failure?: false)
```
Comment (dkubb) [inline spec/unit/mutant/result/subject_spec.rb]: I'd probably vertically align the `neutral_failure?` key here for easier readability.

File: lib/mutant/runner/sink.rb [hunk 2]
```diff
@@ -49,6 +49,20 @@ def result(mutation_result)
 
     private
 
+      # Test if an evil failure has been reported
+      #
+      # @return [Boolean]
+      def evil_failure?
+        !status.subject_results.all?(&:success?)
```
Comment (mbj) [inline lib/mutant/runner/sink.rb]: I'd be okay to have a `FailurePredicate` mixin that defines `#failure?` as the negation of `#success?`. But for consistency we should only ever concretely implement `#success?`.

**Comment by mbj** [comment]
@backus I'm fine with the overall approach but flags.

I do not think the "we terminated because of neutral violation" message has to be added in this itartion.

Terminating on neutral is the correct semantics. Hence this is a free-standing improvement.

To implement the meassaging correct, I'd consider to define a "termination reason" type:

```
data TerminationReason = Done | NeutralViolation Mutation
```

Which than gets expanded by the CLI (or other reporters) to dedicated messaging.

File: lib/mutant/runner/sink.rb (hunk 2 above)
Comment (backus) [inline lib/mutant/runner/sink.rb]: Added the failure predicate. I think it turned out really nice.

File: spec/unit/mutant/result/subject_spec.rb (hunk 1 above)
Comment (backus) [inline spec/unit/mutant/result/subject_spec.rb]: Updated

**Comment by mbj** [comment]
Closing this PR as all contributions will require a CLA from now on.

Also this approach will not fit into the currnent runner. I'll do an implementation of these semantics on my own.

**Pull request closed without merging**